### PR TITLE
Amalgamate multiple CIGAR ops into single entry.

### LINKED ISSUE #1597: bcftools mpileup behaves differently for cigar M vs =/X
I’m working on the read mapper [strobealign](https://github.com/ksahlin/strobealign/). While investigating some false positive SNVs that were called on strobealign’s output but not on BWA-MEM’s, I noticed that it matters whether the CIGAR field in the SAM output uses the `=` and `X` operations or `M` to encode alignments. (Strobealign uses `=`/`X`.)

Here is a small example. The rightmost SNV at position 96305 is the relevant one. The bottommost read has CIGAR `96=1X40=1X11=1X`, and when I change this to `150M`, the SNV disappears.
igv

Corresponding BAM file: [cigar-m-vs-eqx.gz](https://github.com/samtools/htslib/files/11122771/cigar-m-vs-eqx.gz) (please change extension to `.bam`). I was not able to make this any shorter: As soon as I remove any read, the SNV disappears.

The reference is [`GCA_000001405.15_GRCh38_no_alt_analysis_set.fna`](https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.15_GRCh38/seqs_for_alignment_pipelines.ucsc_ids/GCA_000001405.15_GRCh38_no_alt_analysis_set.fna.gz).

These commands will reproduce the problem:
```
bcftools mpileup --fasta-ref GCA_000001405.15_GRCh38_no_alt_analysis_set.fna cigar-m-vs-eqx.bam | grep 96305
samtools view -h cigar-m-vs-eqx.bam | sed 's|96=1X40=1X11=1X|150M|' | samtools view --write-index -o m.bam
bcftools mpileup --fasta-ref GCA_000001405.15_GRCh38_no_alt_analysis_set.fna m.bam | grep 96305
```
Output is
```
chr20   96305   .       C       A,<*>   0       .       DP=1;I16=0,0,0,1,0,0,40,1600,0,0,60,3600,0,0,0,0;QS=0,1,0;SGB=-0.379885;MQ0F=0  PL      40,3,0,40,3,40
```
vs
```
chr20   96305   .       C       <*>     0       .       DP=1;I16=0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0;QS=0,0;MQ0F=0  PL      0,0,0
```
I don’t have an opinion on which of the above is correct, but I would assume the intention is for it to be consistent.

### PR DESCRIPTION
**Pull request by whitwham**
Multiple matching (or sequence (mis)matching)) ops (e.g. 10M40M) give a different VCF using BAQ than a single operation of the same length (e.g. 50M).  This change compresses the multiple operations into one.

In combination with samtools/bcftools#1914 it should fix #1597.

### CODE REVIEW

**Review by jkbonfield — COMMENTED** [review]
Some small corrections, and a suggestion which I can take or leave either way.

It could also do with test data demonstrating the problem, but I can't reproduce it!  I'll take a look at the original issue and see if I can get something smaller to test on.

File: realn.c [hunk 1]
```diff
@@ -268,8 +268,28 @@ int sam_prob_realn(bam1_t *b, const char *ref, hts_pos_t ref_len, int flag) {
...
             uint8_t *rght = tref;
+            int len = 0;
+
             for (k = 0, x = c->pos, y = 0; k < c->n_cigar; ++k) {
                 int op = cigar[k]&0xf, l = cigar[k]>>4;
+
+                // concatinate alignment matches (including sequence (mis)matches)
+                // otherwise 50M50M gives a different result to 100M
+                if (op == BAM_CMATCH || op == BAM_CEQUAL || op == BAM_CDIFF) {
+                    if (((k + 1) < c->n_cigar)) {
```
[the remark this answers is not in the document]
Comment (whitwham) [reply realn.c:279]: I don't remember, I think I had something else in there originally.

**Comment by whitwham** [comment]
> Some small corrections, and a suggestion which I can take or leave either way.
> 
> It could also do with test data demonstrating the problem, but I can't reproduce it! I'll take a look at the original issue and see if I can get something smaller to test on.

To replicate the original error you also need the version of bcftools without the fix.  This change keeps the ID=16 numbers the same, the bcftools one has the alternative bases.

Also, the test data the provided by the users was the smallest they could produce to show the error.

**Comment by jkbonfield** [comment]
I've managed to produce something really tiny now, starting with the original test data provided.  I think that's just simulated data, but I don't know the license on it so will try and figure out the cause of it going wrong to produce our own.

The easiest way I found of reproducing it also, and what we'd need here, is via test/test_realn.  We already have test files, so we can just add one more with two entries using =/X vs M.

**Comment by jkbonfield** [comment]
Here is some test data.

test/realn03.fa
```
>MX
CGTCTACTACG
```

test/realn03.sam
```
@HD	VN:1.6	SO:coordinate
@SQ	SN:MX	LN:11
M=X	163	MX	1	60	11M	*	0	0	CGTCTCCTACG	IIIIIIIIIII
M=X	163	MX	1	60	5=1X5=	*	0	0	CGTCTCCTACG	IIIIIIIIIII
```

Then run:
```
./test/test_realn -e -f test/realn03.fa -i test/realn03.sam
```

You should see the same BQ tag in both cases, but don't with *develop*.  There are already test files for test_realn, so this can just be another 1 to add and put into test.pl

I've verified it's only affecting extended mode, which tallies with the code changes.

**Pull request merged**